### PR TITLE
Overview of the symbol layer types

### LINKED ISSUE #710: [FEATURE] Add geometry generator symbols
Original commit: https://github.com/qgis/QGIS/commit/17f9d554f8a4375ae06af40521ffa3a012eac331 by m-kuhn

Unfortunately this lazy coder did not write a description... :-(

### LINKED ISSUE #947: [FEATURE] add outline join style settings to simple markers
(fixes #3797)
Original commit: https://github.com/qgis/QGIS/commit/6ad353740399ac752da83a6d6d6c23fd3740d286 by nirvn

Unfortunately this naughty coder did not write a description... :-(

### LINKED ISSUE #951: [FEATURE] add outline join style to ellipse markers
Original commit: https://github.com/qgis/QGIS/commit/541463718d2863ca92336e40a4f74eb62da6a0e5 by nyalldawson

Unfortunately this naughty coder did not write a description... :-(

### LINKED ISSUE #977: [Feature] Add arrow symbol layer
Original commit: https://github.com/qgis/QGIS/commit/e849cad11d491911fe178f8a2cd296e5cb9fab90 by mhugo
Merge pull request #2955 from mhugo/arrows

[Feature] Arrow symbol layer

### LINKED ISSUE #984: [FEATURE] add outline settings to font markers
Original commit: https://github.com/qgis/QGIS/commit/9735c13e33ac79a201b691b19178a7e4dbff5b33 by nirvn

Unfortunately this naughty coder did not write a description... :-(

### LINKED ISSUE #998: [FEATURE] add setting to control centroid fill point rendering
on all parts or a single part of multi-features (fixes #9199)
Original commit: https://github.com/qgis/QGIS/commit/fef8c1fef2ad60bf6f858b2253e57c463f40659e by nirvn

Unfortunately this naughty coder did not write a description... :-(

### LINKED ISSUE #1015: [FEATURE] New "Filled marker" symbol layer type
Original commit: https://github.com/qgis/QGIS/commit/ea2014923914d21e5414b33461c8b40fe39f92dd by nyalldawson

A "filled marker" is similar to the simple marker symbol layer, except
that it uses a fill sub symbol to render the marker. This allows use
of all the existing QGIS fill (and outline) styles for rendering
markers, eg gradient or shapeburst fills.

A new abstract class QgsSimpleMarkerSymbolLayerBase has been created
from which both QgsSimpleMarkerSymbolLayerV2 and
QgsFilledMarkerSymbolLayer derive. All common code has been moved to
this base class, allowing for easy addition of new shapes or features
to both simple markers and filled markers without duplication.

Also, adds a lot of missing docs to marker symbol layers.

### PR DESCRIPTION
**Pull request by DelazJ**
Fixes #710, fixes #947, fixes #951, fixes #977, fixes #984, fixes #998, fixes #1015

### CODE REVIEW

File: source/docs/user_manual/working_with_vector/style_library.rst [hunk 1]
```diff
@@ -298,131 +298,218 @@ The symbol layer
...
+
+Available with all type of symbols, the :guilabel:`geometry generator` symbol
+layer allows to use expression syntax to generate a geometry on the fly during
+the rendering process. The resulting geometry does not have to match with the
+original geometry type and you can add several differently modified symbol
+layers on top of each other.
+
+Some examples:
+
+.. code-block:: sql
```
Comment (SrNetoChan) [inline source/docs/user_manual/working_with_vector/style_library.rst]: Should we use a Python code-block and use # to comment? AFAIK Expression builder comments are done with #.

Not important, just an idea.

Comment (DelazJ) [inline source/docs/user_manual/working_with_vector/style_library.rst]: Actually comments are done with `--` and not `#` in the expression builder, reason why I used it (someone may copy-paste the code with comments).
But I have an issue with the rendering of this code sample (displays color in github but monocolor in html).

Comment (SrNetoChan) [inline source/docs/user_manual/working_with_vector/style_library.rst]: Sorry for the noise then.

File: source/docs/user_manual/working_with_vector/style_library.rst [hunk 2]
```diff
@@ -298,131 +298,218 @@ The symbol layer
...
+Appropriate for point geometry features, marker symbols have several
+:guilabel:`Symbol layer types`:
+
+* **Simple marker** (default)
+* **Ellipse marker**: a simple marker symbol layer, with customizable width and
+  height
+* **Filled marker**: similar to the simple marker symbol layer, except that it
+  uses a :ref:`fill sub symbol <vector_fill_symbols>` to render the marker.
+  This allows use of all the existing QGIS fill (and outline) styles for
+  rendering markers, e.g. gradient or shapeburst fills.
```
Comment (SrNetoChan) [inline source/docs/user_manual/working_with_vector/style_library.rst]: This topic ends with a (.) period, while others don't. I don't have a strong oppinion on what is best, but can we always use the same? (I personally like semi-colons :-))

File: source/docs/user_manual/working_with_vector/style_library.rst [hunk 3]
```diff
@@ -298,131 +298,218 @@ The symbol layer
...
 * :guilabel:`Size`
 * :guilabel:`Outline style`
 * :guilabel:`Outline width`
-* :guilabel:`Angle`
+* :guilabel:`Join Style`
+* :guilabel:`Rotation`
 * :guilabel:`Offset X,Y`: You can shift the symbol in the x- or y-direction.
 * :guilabel:`Anchor point`
-* :guilabel:`Data defined properties ...`
+* a preview of predefined symbols to choose from
```
Comment (SrNetoChan) [inline source/docs/user_manual/working_with_vector/style_library.rst]: Is this in the right place?

File: source/docs/user_manual/working_with_vector/style_library.rst [hunk 4]
```diff
@@ -298,131 +298,218 @@ The symbol layer
...
-only inside polygon`.
-
-**Note:** When geometry type is polygon, you can choose to disable the
-automatic clipping of lines/polygons to the canvas extent. In
-some cases this clipping results in unfavourable symbology (eg
-centroid fills where the centroid must always be the actual
-feature's centroid).
-
+.. note:: When geometry type is polygon, you can choose to disable the
+ automatic clipping of lines/polygons to the canvas extent. In
```
Comment (SrNetoChan) [inline source/docs/user_manual/working_with_vector/style_library.rst]: Bad indent in the lines following this note

File: source/docs/user_manual/working_with_vector/style_library.rst (hunk 3 above)
Comment (DelazJ) [inline source/docs/user_manual/working_with_vector/style_library.rst]: There are two kinds of symbols preview:
* one described line 275, at the symbol level, which shows "real" symbols from the Symbol manager
* and this one, at the marker symbol layer level (available for all but geometry generator, arrow and vector field)

File: source/docs/user_manual/working_with_vector/style_library.rst [hunk 5]
```diff
@@ -298,131 +298,218 @@ The symbol layer
...
+
+* **Simple marker** (default)
+* **Ellipse marker**: a simple marker symbol layer, with customizable width and
+  height
+* **Filled marker**: similar to the simple marker symbol layer, except that it
+  uses a :ref:`fill sub symbol <vector_fill_symbols>` to render the marker.
+  This allows use of all the existing QGIS fill (and outline) styles for
+  rendering markers, e.g. gradient or shapeburst fills.
+* **Font marker**: use installed fonts as marker symbols
+* **Geometry generator** (see geometry_generator_symbol_)
```
Comment (SrNetoChan) [inline source/docs/user_manual/working_with_vector/style_library.rst]: Can we use the :ref: role in here to render the section's title?

File: source/docs/user_manual/working_with_vector/style_library.rst [hunk 6]
```diff
@@ -298,131 +298,218 @@ The symbol layer
...
+* **SVG marker**: provides you with images from your SVG paths (set in
+  :menuselection:`Settings --> Options --> System` menu) to render as marker symbol.
+  Each SVG file colors and outline can be adapted.
+* **Vector Field marker** (see vector_field_marker_)
```
Comment (SrNetoChan) [inline source/docs/user_manual/working_with_vector/style_library.rst]: Can we use the :ref: role in here to render the section's title?

File: source/docs/user_manual/working_with_vector/style_library.rst [hunk 7]
```diff
@@ -298,131 +298,218 @@ The symbol layer
...
+
+.. _arrow_symbol:
+
+* **Arrow**: draw lines as curved (or not) arrows with a single or a double
+  head with configurable width, length and thickness. It also uses a
+  :ref:`fill sub symbol <vector_fill_symbols>` such as gradients or shapeburst
+  to render the arrow body.
+  Combined with the geometry generator, this type of layer symbol helps you
+  representing flow maps.
+* **Geometry generator** (see geometry_generator_symbol_)
```
Comment (SrNetoChan) [inline source/docs/user_manual/working_with_vector/style_library.rst]: Can we use the :ref: role in here to render the section's title?

File: source/docs/user_manual/working_with_vector/style_library.rst [hunk 8]
```diff
@@ -298,131 +298,218 @@ The symbol layer
...
+^^^^^^^^^^^^
+
+Appropriate for polygon geometry features, fill symbols have also several
+symbol layer types:
+
+* **Simple fill** (default)
+* **Centroid fill**: places a marker symbol at the centroid of the visible feature.
+  The marker can be placed on every part of a multi-part feature or only on its
+  biggest part, and forced to be inside the polygon.
+* **Geometry generator** (see geometry_generator_symbol_)
```
Comment (SrNetoChan) [inline source/docs/user_manual/working_with_vector/style_library.rst]: Can we use the :ref: role in here to render the section's title?

File: source/docs/user_manual/working_with_vector/style_library.rst (hunk 4 above)
Comment (SrNetoChan) [inline source/docs/user_manual/working_with_vector/style_library.rst]: Never mind, it works well as it's following the first line. A always prefer to put the note in the next line.

```
.. note::

   When geometry type is polygon, you can choose to disable the
   automatic clipping of lines/polygons to the canvas extent. In (...)
```

But it's just a question of taste.

File: source/docs/user_manual/working_with_vector/style_library.rst (hunk 3 above)
Comment (SrNetoChan) [inline source/docs/user_manual/working_with_vector/style_library.rst]: It's just that the list is for properties you can set. I would make reference to this in a new sentence. No strong feeling about it anyway.

**Comment by SrNetoChan** [comment]
Apart from some comments, LGTM. Thanks @DelazJ !

File: source/docs/user_manual/working_with_vector/style_library.rst (hunk 4 above)
Comment (DelazJ) [inline source/docs/user_manual/working_with_vector/style_library.rst]: I do like the second too (especially for multiline comments like this one). The one I used however saves one (or two) lines in the stats :)

**Comment by malleechick** [comment]
I'm happy to start on the images that go with this #1738. 
I'm on a Windows 10 machine, so my screen shots will be different from the standard. Is that a problem?
I also see that the text says the dynamic symbol indicator is on the left of the dialogue window. On mine it is on the right. Do we need to specify if we have images?

**Comment by SrNetoChan** [comment]
@malleechick screenshots of the GUI we do using ubuntu, we would have to replace them later. any chance you can use a virtual machine with ubuntu for that?

Anyway, just thinking of examples and creating them on windows would be a good help. Just don't close the issue after.

**Comment by DelazJ** [comment]
@SrNetoChan @malleechick I just notice your discussion. I was rather thinking of samples in the spirit of the one added at http://docs.qgis.org/testing/en/docs/user_manual/working_with_vector/vector_properties.html#placement-for-polygon-layers (meaning no GUI items (hence no OS issue), just feature geometries with different symbol types applied to it). I don't know if my explanation is clear but if you think we need to show the symbol layers too, let's do it.

**Comment by SrNetoChan** [comment]
Ohhh. In that case, the OS doesn't really matter.

**Comment by DelazJ** [comment]
btw I fixed the issues; is it ok now (I tried to follow the `;` rules but i haven't been convinced in all places :-) )?

**Comment by malleechick** [comment]
I'm happy to start on the images that go with this #1738. 
I'm on a Windows 10 machine, so my screen shots will be different from the standard. Is that a problem?
I also see that the text says the dynamic symbol indicator is on the left of the dialogue window. On mine it is on the right. Do we need to specify if we have images?

**Comment by DelazJ** [comment]
> I'm on a Windows 10 machine, so my screen shots will be different from the standard. Is that a problem?

no. But give a look to my comment above
>I also see that the text says the dynamic symbol indicator is on the left of the dialogue window. On mine it is on the right. 

I think this kind of information has been removed in some recent commit. Can you point to the text you are refering to?

**Pull request merged**